gems/sorbet-runtime/lib/types/private/abstract/declare.rb

```diff
@@ -27,26 +27,26 @@ def self.declare_abstract(mod, type:)
         raise "Classes can't be interfaces. Use `abstract!` instead of `interface!`."
       end
 
-      if mod.instance_method(:initialize).owner == mod
-        raise "You must call `abstract!` *before* defining an initialize method"
+      if mod.method(:new).owner == mod
+        raise "You must call `abstract!` *before* defining a `new` method"
```

Comment:
WDYT about also warning people that you shouldn't define `new` on `abstract!` classes?

Reply:
It's interesting. I think it's probably a good check to have? I'm struggling to think of a case where you'd want to define `self.new` and not `initialize`.

But also I'm generally afraid of rocking the boat too much on runtime changes like this, because people do weird things.

So I'd probably prefer to investigate something like raising an error for any `self.new` method defined in an abstract class in a follow up change. Or alternatively, if we do just want to emit a warning, not an error, we could get rid of the `alias_method` trick we do later in the file, which would then use the Ruby VM's built in method redefinition warnings to surface this to the user.

Reply:
https://github.com/rails/rails/blob/ef04fbb3b256beececfa44c47c4ec93ac6945e59/activerecord/lib/active_record/inheritance.rb#L56-L78

Rails defines `self.new` in order to implement single-table inheritance. `new` is called on the STI base class, which delegates to the appropriate child class based on the value of the `type` attribute.

Consider this example:
- `Market.has_many :regions`
- `Region < ActiveRecord::Base` and is marked `abstract!`
- `Country < Region`
- `Province < Region`

Before this change (desired behavior IMO)
```
[1] pry(main)> market.regions.build
RuntimeError: Region is declared as abstract; it cannot be instantiated
[2] pry(main)> market.regions.build(type: Country)
=> #<Country:0x00007f97c3b1a3c0 ... >
```

After this change:
```
[1] pry(main)> market.regions.build
RuntimeError: Region is declared as abstract; it cannot be instantiated
[2] pry(main)> market.regions.build(type: Country)
RuntimeError: Region is declared as abstract; it cannot be instantiated
```

Reply:
@cbothner https://github.com/sorbet/sorbet/pull/7079